### ISSUE TITLE
Can't disconnect unless you are actually connected to server.

### ISSUE THREAD

**Issue by thbaja** [daltoniam/Starscream#127]
Using websockets in real life conditions, I sometimes experience that the connection is lost without Starscream knowing about. This is probably due to firewalls etc., even though I try to mitigate it using SSL.

So I ping the server at regular intervals, and if I don't receive a pong within a given period I will disconnect and try to establish a new connection.
To simulate these conditions I use apple's network conditioner, and if it set packets dropped to 100% I cannot disconnect with the socket.disconnect() method. At least not until I let packets flowing again.

Without having looked more into it, I suspect that we end up waiting here:
https://github.com/daltoniam/Starscream/blob/master/WebSocket.swift#L688

And in general it might be a good idea to have a timeout on that write.

Sorry for not giving more in-depth feedback. Maybe I get a chance to look at it over the next weekend.

**Comment by thbaja** [comment]
After about 70 seconds of not being able to connect, the error delegate handler will give this (even after having had disconnect() being called): 
Error Domain=NSPOSIXErrorDomain Code=60 "The operation couldn’t be completed. Operation timed out"

**Comment by daltoniam** [comment]
Yeah I can see how the might be difficult to work with. From what you are describing, that sounds like normal socket behavior. The socket isn't disconnected, all the packets are just being dropped, so it keeps trying to resend them until the socket gives up after the 60 or so seconds. I added add an option that I think might help in this case.
https://github.com/daltoniam/Starscream/blob/master/WebSocket.swift#L142

Basically it will forcible disconnect the socket by closing the streams.

```swift
socket.disconnect(5) //forcible disconnects after 5 seconds
```

`disconnect()` with no parameters continues to work the way it did.

**Comment by thbaja** [comment]
Thanks for looking into this.

However this disconnect makes it impossible to reconnect using the same socket. So a re-instantiation of the socket object is necessary.

**Comment by daltoniam** [comment]
Interesting, not sure why that would be, it is using the same `disconnectStream` method as all the other disconnect logic does. Does it report why it is unable to connect again or do you see a point in the code it is failing?
